### PR TITLE
Task/g357 standardize transaction service

### PR DESCRIPTION
**Pull request by rg911**
- Added service to resolve alias(es) for transactions from block receipts. 
- Fixed unresolvedAddress bug in `ResolutionStatement`
- Added e2e tests

**Please note that there's discussion on NEM2 slack channel regarding the `PrimaryId` value in `ResolutionStatement` as if `Harvest_Fee` receipt exists on the block, the transaction index is less `1` against the `Resolution entry primaryId`. Solution here may change upon investigation results on catapult-server**

Issue: #357

### CODE REVIEW

File: src/service/TransactionService.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+     * Constructor
+     * @param url Base catapult-rest url
+     */
+    constructor(url: string) {
+        this.transactionHttp = new TransactionHttp(url);
+        this.receiptHttp = new ReceiptHttp(url);
+    }
+
+    /**
+     * @internal
```
Comment (evias) [inline src/service/TransactionService.ts]: could this method be exposed outside too?

File: src/service/TransactionService.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        const resolutionStatement = (resolutionType === ResolutionType.Address ? statement.addressResolutionStatements :
+            statement.mosaicResolutionStatements).find((resolution) => resolution.height.toString() === height &&
+                (resolution.unresolved as NamespaceId).equals(unresolved));
+
+        if (!resolutionStatement) {
+            throw new Error('No resolution statement found');
+        }
+        // source (0,0) is reserved for blocks, source (n, 0) is for txes, where n is 1-based index
+        const resolutionEntry = resolutionStatement.resolutionEntries
+            .find((entry) => entry.source.primaryId ===
```
Comment (dgarcia360) [inline src/service/TransactionService.ts]: ``entry.source.primaryId ===``  might not work always if there is more than one transaction in the block using the same alias (and the alias does not change within the block). I believe there is only one resolution statement if this is repeated.

**Example**
_Block:_
Height: 5
Transactions:
* 1 - Bob -> 1 random.token -> Alice
* 2- Carol -> 1 random.token > Denis

_Receipts for 5_
* mosaicResolutionStatement (random.token -> ID starting at 1)

File: src/service/TransactionService.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+     * Extract resolved address | mosaic from block receipt
+     * @param resolutionType Resolution type: Address / Mosaic
+     * @param unresolved Unresolved address / mosaicId
+     * @param statement Block receipt statement
+     * @param transactionIndex Transaction index
+     * @param height Transaction height
+     * @param aggregateTransactionIndex Transaction index for aggregate
+     * @returns {MosaicId | Address}
+     */
+    public static getResolvedFromReceipt(resolutionType: ResolutionType,
```
Comment (fboucquez) [inline src/service/TransactionService.ts]: I'll split this method in 2, getResolvedAddressFromRecipient and getResolvedMosaicIdFromRecipient

File: src/model/transaction/AccountAddressRestrictionTransaction.ts [hunk 4]
```diff
@@ -190,4 +195,59 @@ export class AccountAddressRestrictionTransaction extends Transaction {
...
         return transactionBuilder.serialize();
     }
+
+    /**
+     * @internal
+     * @param receiptHttp ReceiptHttp
+     * @param aggregateTransactionIndex Transaction index for aggregated transaction
+     * @returns {Observable<AccountAddressRestrictionTransaction>}
+     */
+    resolveAliases(receiptHttp: ReceiptHttp, aggregateTransactionIndex?: number): Observable<AccountAddressRestrictionTransaction> {
```
Comment (fboucquez) [inline src/model/transaction/AccountAddressRestrictionTransaction.ts]: I am not fan of the double dispatch. It gets the Transaction model too complicated with internal methods the user doesn't really care. I would create a switch/handers statement in the TransactionService (especially when there is a static call  TransactionService.getResolvedFromReceipt) and remove resolveAliases from the transaction hierarchy.

I would stop using the Http implementations and start using the Repository Interface, in this case, ReceiptRepository

Comment (rg911) [reply src/model/transaction/AccountAddressRestrictionTransaction.ts]: Do you mean to put all codes in to one class here? It's another option agree but the single class  can be complicated and main thing here is you will lose the single responsibility for Transaction to resolve alias. 

Regrading the interface, ts doesn't work like JAVA or C# which you can use unity or spring, There's no DI in TS SDK, so  resolving the interface is not easy

File: src/service/TransactionService.ts (hunk 3 above)
Comment (rg911) [reply src/service/TransactionService.ts]: Can do but will probably create blocks of duplicated codes ??

Comment (fboucquez) [reply src/service/TransactionService.ts]: you can have a third method that does the filtering primary/secondary filtering. I'll push the java example

File: src/model/transaction/AggregateTransaction.ts [hunk 5]
```diff
@@ -400,4 +403,32 @@ export class AggregateTransaction extends Transaction {
...
+     * @param receiptHttp ReceiptHttp
+     * @returns {Observable<AggregateTransaction>}
+     */
+    resolveAliases(receiptHttp: ReceiptHttp): Observable<AggregateTransaction> {
+        const transactionInfo = this.checkTransactionHeightAndIndex();
+        return from(this.innerTransactions).pipe(
+            mergeMap((transaction) => transaction.resolveAliases(receiptHttp, transactionInfo.index)),
+            map((transaction) => transaction as InnerTransaction),
+            toArray(),
+        ).pipe(
```
Comment (fboucquez) [inline src/model/transaction/AggregateTransaction.ts]: Should aggregateTransactionIndex be provided to resolve aliases here? aggregateTransactionIndex seems like it's never provided.

File: src/model/transaction/AccountAddressRestrictionTransaction.ts (hunk 4 above)
Comment (fboucquez) [reply src/model/transaction/AccountAddressRestrictionTransaction.ts]: Not about DI, it's just about using interfaces, you can do

`resolveAliases(recipientRepository: RecipientRepository, aggregateTransactionIndex?: number)`

File: src/model/receipt/ResolutionStatement.ts [hunk 6]
```diff
@@ -102,38 +102,70 @@ export class ResolutionStatement {
...
-            /*
-            If no most recent secondaryId matched transaction index, find previous resolution entry (most recent).
-            */
+        /*
+        If no most recent secondaryId matched transaction index, find previous resolution entry (most recent).
+        This means the resolution entry for the specific innter transaction (inside Aggregate) /
+        was generated previously outside the aggregate. It should return the previous entry (privious primarId)
+
+        e.g. Given:
+        Entries: [{P:1, S:0}, {P:2, S:0}, {P5:, S:6}]
```
Comment (dgarcia360) [inline src/model/receipt/ResolutionStatement.ts]: P5: --> P:5

File: src/model/receipt/ResolutionStatement.ts [hunk 7]
```diff
@@ -102,38 +102,70 @@ export class ResolutionStatement {
...
+        // When transaction index matches a primary id, get the most recent secondaryId (resolvedPrimaryId can only <= primaryId)
+        const resolvedSecondaryId = this.getMaxSecondaryIdByPrimaryIdAndSecondaryId(resolvedPrimaryId, secondaryId);
 
-        if (resolvedSecondaryId === 0 && secondaryId !== resolvedSecondaryId) {
-            /*
-            If no most recent secondaryId matched transaction index, find previous resolution entry (most recent).
-            */
+        /*
+        If no most recent secondaryId matched transaction index, find previous resolution entry (most recent).
+        This means the resolution entry for the specific innter transaction (inside Aggregate) /
```
Comment (dgarcia360) [inline src/model/receipt/ResolutionStatement.ts]: inter --> inner

File: src/model/receipt/ResolutionStatement.ts [hunk 8]
```diff
@@ -102,38 +102,70 @@ export class ResolutionStatement {
...
+            return this.resolutionEntries.find((entry) => entry.source.primaryId === lastPrimaryId &&
+                entry.source.secondaryId === this.getMaxSecondaryIdByPrimaryId(lastPrimaryId));
         }
-        // All matched
+
+        /*
+        Found a matched resolution entry on both primaryId and secondaryId
+
+        e.g. Given:
+        Entries: [{P:1, S:0}, {P:2, S:0}, {P5:, S:6}]
```
Comment (dgarcia360) [inline src/model/receipt/ResolutionStatement.ts]: P5: --> P:5

File: src/model/receipt/ResolutionStatement.ts [hunk 9]
```diff
@@ -102,38 +102,70 @@ export class ResolutionStatement {
         */
         const resolvedPrimaryId = this.getMaxAvailablePrimaryId(primaryId);
 
+        /*
+        If no primaryId found, it means there's no resolution entry available for the process. Invalid entry.
+
+        e.g. Given:
+        Entries: [{P:2, S:0}, {P5:, S:6}]
```
Comment (dgarcia360) [inline src/model/receipt/ResolutionStatement.ts]: P5: -> P:5

File: src/model/receipt/ResolutionStatement.ts [hunk 10]
```diff
@@ -102,38 +102,70 @@ export class ResolutionStatement {
...
+        */
         if (resolvedPrimaryId === 0) {
             return undefined;
         } else if (primaryId > resolvedPrimaryId) {
             /*
             If the transaction index is greater than the overall most recent source primary id.
             Use the most recent resolution entry (Max.PrimaryId + Max.SecondaryId)
+
+            e.g. Given:
+            Entries: [{P:1, S:0}, {P:2, S:0}, {P:4, S:2}, {P:4, S:4} {P7:, S:6}]
```
Comment (dgarcia360) [inline src/model/receipt/ResolutionStatement.ts]: P:7

File: src/model/receipt/ResolutionStatement.ts [hunk 11]
```diff
@@ -102,38 +102,70 @@ export class ResolutionStatement {
...
+            It should return Entry: {P:2, S:0}
             */
             return this.resolutionEntries
                 .find((entry) => entry.source.primaryId === resolvedPrimaryId &&
                     entry.source.secondaryId === this.getMaxSecondaryIdByPrimaryId(resolvedPrimaryId));
         }
 
-        // When transaction index matches a primary id, get the most recent secondaryId
-        const resolvedSecondaryId = this.getMaxSecondaryIdByPrimaryId(resolvedPrimaryId);
+        // When transaction index matches a primary id, get the most recent secondaryId (resolvedPrimaryId can only <= primaryId)
```
Comment (dgarcia360) [inline src/model/receipt/ResolutionStatement.ts]: primary id -> primaryId

File: src/model/receipt/ResolutionStatement.ts (hunk 7 above)
Comment (dgarcia360) [inline src/model/receipt/ResolutionStatement.ts]: innter -> inner

File: src/model/receipt/ResolutionStatement.ts [hunk 12]
```diff
@@ -102,38 +102,70 @@ export class ResolutionStatement {
...
+        const resolvedSecondaryId = this.getMaxSecondaryIdByPrimaryIdAndSecondaryId(resolvedPrimaryId, secondaryId);
 
-        if (resolvedSecondaryId === 0 && secondaryId !== resolvedSecondaryId) {
-            /*
-            If no most recent secondaryId matched transaction index, find previous resolution entry (most recent).
-            */
+        /*
+        If no most recent secondaryId matched transaction index, find previous resolution entry (most recent).
+        This means the resolution entry for the specific innter transaction (inside Aggregate) /
+        was generated previously outside the aggregate. It should return the previous entry (privious primarId)
```
Comment (dgarcia360) [inline src/model/receipt/ResolutionStatement.ts]: privious -> previous
PrimarId -> primaryId

File: src/model/receipt/ResolutionStatement.ts [hunk 13]
```diff
@@ -102,38 +102,70 @@ export class ResolutionStatement {
...
             If the transaction index is greater than the overall most recent source primary id.
             Use the most recent resolution entry (Max.PrimaryId + Max.SecondaryId)
+
+            e.g. Given:
+            Entries: [{P:1, S:0}, {P:2, S:0}, {P:4, S:2}, {P:4, S:4} {P7:, S:6}]
+            Transaction: [Inx:5(0+1), AggInx:0]
+            It should return Entry: {P:3, S:4}
+
+            e.g. Given:
+            Entries: [{P:1, S:0}, {P:2, S:0}, {P:4, S:2}, {P:4, S:4}, {P7:, S:6}]
```
Comment (dgarcia360) [inline src/model/receipt/ResolutionStatement.ts]: P7: -> P:7

File: src/model/receipt/ResolutionStatement.ts (hunk 8 above)
Comment (dgarcia360) [inline src/model/receipt/ResolutionStatement.ts]: P5->P:5

File: src/model/transaction/AccountAddressRestrictionTransaction.ts (hunk 4 above)
Comment (rg911) [reply src/model/transaction/AccountAddressRestrictionTransaction.ts]: How do you find the correct implementation of the injected interface?

File: src/model/transaction/AggregateTransaction.ts (hunk 5 above)
Comment (rg911) [reply src/model/transaction/AggregateTransaction.ts]: outdated

File: src/service/TransactionService.ts (hunk 3 above)
Comment (rg911) [reply src/service/TransactionService.ts]: outdated

**Pull request merged**